{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from bokeh.plotting import figure, output_notebook, show, gridplot\n",
    "from bokeh.models.tickers import FixedTicker\n",
    "from bokeh.models import NumeralTickFormatter, Range1d, LinearAxis, ColumnDataSource, HoverTool\n",
    "\n",
    "import sys\n",
    "import os\n",
    "import pyfasta\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import allel\n",
    "import functools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "class GenomePlot:\n",
    "    \n",
    "    genome = None\n",
    "    contigs = None\n",
    "    tools = \"pan,wheel_zoom,box_zoom,save,reset\"\n",
    "    layout = None\n",
    "    pfunc = None\n",
    "    min_border_left = 50\n",
    "    min_border_right = 10\n",
    "    major_tick_dist = 1e7\n",
    "    plot_width_per_mb = 6\n",
    "    chrom_label_func = lambda self, y: y\n",
    "    \n",
    "    def __init__(self, fasta, contigs=None, layout=None, share_y=True, pfunc=None):\n",
    "        \n",
    "        self.genome = pyfasta.Fasta(fasta)\n",
    "        \n",
    "        if contigs is None:\n",
    "            self.contigs = list(self.genome.keys())\n",
    "        else:\n",
    "            self.contigs = contigs\n",
    "            \n",
    "        # handle layout\n",
    "        if layout is not None:\n",
    "            self.layout = self.parse_layout(layout)\n",
    "        else:\n",
    "            self.layout = self.auto_layout()\n",
    "        \n",
    "        if pfunc is not None:\n",
    "            self.pfunc = pfunc\n",
    "                \n",
    "    def parse_layout(self, lstring):\n",
    "        layout = list()\n",
    "        assert type(lstring) == str, \"layout must be a string\"\n",
    "        q = list(self.contigs)\n",
    "        q.reverse()\n",
    "            \n",
    "        rows = lstring.split(\"|\")\n",
    "        for row in rows:\n",
    "            r = list()\n",
    "            for x in row:\n",
    "                assert x in \"ox.\", \"Only chars o, x, . are allowed\"\n",
    "                if x == \".\":\n",
    "                    r.append(None)\n",
    "                else:\n",
    "                    r.append(q.pop())\n",
    "            layout.append(r)\n",
    "        return layout\n",
    "    \n",
    "    def auto_layout():\n",
    "        pass\n",
    "\n",
    "    def apply(self, func, **kwargs):\n",
    "        \n",
    "        # create a figure with specified layout        \n",
    "        d = [[] for i in range(len(self.layout))]\n",
    "        \n",
    "        for i, row in enumerate(self.layout):\n",
    "            \n",
    "            is_bottom = ((i + 1) == len(self.layout))\n",
    "            \n",
    "            for j, chrom in enumerate(row):\n",
    "                \n",
    "                is_left = (j == 0)\n",
    "                \n",
    "                if chrom is not None:\n",
    "                    \n",
    "                    csize = len(self.genome[chrom])\n",
    "                    px = int(csize * 1e-6 * self.plot_width_per_mb) + self.min_border_left + self.min_border_right\n",
    "                    \n",
    "                    try:\n",
    "                        yrange = s1.y_range\n",
    "                    except NameError:\n",
    "                        yrange = None\n",
    "\n",
    "                    s1 = figure(width=px, \n",
    "                                min_border_left=self.min_border_left,\n",
    "                                min_border_right=self.min_border_right,\n",
    "                                plot_height=250, \n",
    "                                tools=self.tools,\n",
    "                                title=self.chrom_label_func(chrom),\n",
    "                                y_range=yrange,\n",
    "                                x_range=(1, csize))\n",
    "\n",
    "                    s1.xaxis.ticker = FixedTicker(ticks=np.arange(0, csize, self.major_tick_dist))\n",
    "                    s1.xaxis[0].formatter = NumeralTickFormatter(format=\"0a.0\")\n",
    "                    \n",
    "                    # handle general plot things specific to genome not data\n",
    "                    if self.pfunc is not None:\n",
    "                        self.pfunc(chrom, s1)\n",
    "\n",
    "                    # function lives here\n",
    "                    func(chrom, s1, is_left=is_left, is_bottom=is_bottom, **kwargs)\n",
    "                    d[i].append(s1)\n",
    "\n",
    "                else:\n",
    "                    d[i].append(None)\n",
    "            \n",
    "            \n",
    "        # NEW: put the subplots in a gridplot\n",
    "        p = gridplot(d, toolbar_location=\"left\", sizing_mode='fixed', plot_width=None)\n",
    "        \n",
    "        show(p)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
